File: metacasa-ios/MetaCasa/Models/Budget.swift
import Foundation

struct BudgetPeriod: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var periodType: PeriodType
    var periodStart: Date
    var periodEnd: Date
    var totalIncome: Decimal
    var totalAllocated: Decimal
    var readyToAssign: Decimal
    var notes: String?
    let createdAt: Date?
    var updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case periodType = "period_type"
        case periodStart = "period_start"
        case periodEnd = "period_end"
        case totalIncome = "total_income"
        case totalAllocated = "total_allocated"
        case readyToAssign = "ready_to_assign"
        case notes
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

enum PeriodType: String, Codable, Hashable, Sendable, CaseIterable {
    case week, biweek, month, quarter, year, custom

    var label: String {
        switch self {
        case .week: "Semanal"
        case .biweek: "Quincenal"
        case .month: "Mensual"
        case .quarter: "Trimestral"
        case .year: "Anual"
        case .custom: "Personalizado"
        }
    }
}

struct BudgetAllocation: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var periodId: UUID
    var category: String
    var subcategory: String
    var allocated: Decimal
    var rolloverFromPrev: Decimal
    var rolloverMode: RolloverMode
    var currency: String
    let createdAt: Date?
    var updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case periodId = "period_id"
        case category
        case subcategory
        case allocated
        case rolloverFromPrev = "rollover_from_prev"
        case rolloverMode = "rollover_mode"
        case currency
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

enum RolloverMode: String, Codable, Hashable, Sendable, CaseIterable {
    case none, surplus, full

    var label: String {
        switch self {
        case .none: "Sin rollover"
        case .surplus: "Solo sobrante"
        case .full: "Todo el saldo"
        }
    }
}

/// Cálculo ligero del saldo de un envelope en cliente.
/// La fuente canónica sigue siendo la función SQL `public.envelope_balance(...)`.
struct EnvelopeStatus: Hashable, Sendable {
    let category: String
    let subcategory: String
    let allocated: Decimal
    let spent: Decimal

    var remaining: Decimal { allocated - spent }
    var percentUsed: Double {
        guard allocated > 0 else { return 0 }
        return min(1, max(0, (spent as NSDecimalNumber).doubleValue / (allocated as NSDecimalNumber).doubleValue))
    }
    var isOverBudget: Bool { spent > allocated }
}


/// Resumen canónico de un período, tal como lo devuelve `public.budget_period_summary`.
///
/// Es la ÚNICA definición de "listo para asignar" que debe leer la app. Las tres cifras vienen
/// ya convertidas a la moneda base del hogar.
struct BudgetPeriodSummary: Decodable, Sendable {
    let periodId: UUID
    let baseCurrency: String
    let totalIncome: Decimal
    /// Suma de `allocated`, SIN el arrastre. Es lo que se descuenta del ingreso.
    let totalAllocated: Decimal
    /// Suma de `allocated + rollover_from_prev`. Es lo que hay en los sobres.
    ///
    /// El arrastre no se resta del ingreso: se fondeó con el ingreso del mes ANTERIOR, y
    /// descontarlo de éste sería cobrarlo dos veces. Es exactamente la divergencia que se activó
    /// el día que se arregló el rollover.
    let totalBudgeted: Decimal
    let readyToAssign: Decimal
    /// Sobres cuya moneda no tiene cotización. No se asumen en 1: se cuentan aparte para poder
    /// avisar que el número está incompleto en vez de mostrarlo como si fuera exacto.
    let fxMissingCount: Int

    enum CodingKeys: String, CodingKey {
        case periodId = "period_id"
        case baseCurrency = "base_currency"
        case totalIncome = "total_income"
        case totalAllocated = "total_allocated"
        case totalBudgeted = "total_budgeted"
        case readyToAssign = "ready_to_assign"
        case fxMissingCount = "fx_missing_count"
    }
}

/// Es el default de la columna en Postgres.
extension PeriodType: UnknownTolerantDecodable {
    static var unknownFallback: Self { .month }
}

/// Lo conservador: ante la duda no arrastra saldo, que sería inventar plata.
extension RolloverMode: UnknownTolerantDecodable {
    static var unknownFallback: Self { .none }
}

File: metacasa-flutter/lib/models/budget.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';

part 'budget.freezed.dart';
part 'budget.g.dart';

/// Tipo de período presupuestario.
enum PeriodType {
  @JsonValue('week')
  week,
  @JsonValue('biweek')
  biweek,
  @JsonValue('month')
  month,
  @JsonValue('quarter')
  quarter,
  @JsonValue('year')
  year,
  @JsonValue('custom')
  custom;

  String get label => switch (this) {
        PeriodType.week => 'Semanal',
        PeriodType.biweek => 'Quincenal',
        PeriodType.month => 'Mensual',
        PeriodType.quarter => 'Trimestral',
        PeriodType.year => 'Anual',
        PeriodType.custom => 'Personalizado',
      };
}

/// Modo de rollover de un envelope entre períodos.
enum RolloverMode {
  @JsonValue('none')
  none,
  @JsonValue('surplus')
  surplus,
  @JsonValue('full')
  full;

  String get label => switch (this) {
        RolloverMode.none => 'Sin rollover',
        RolloverMode.surplus => 'Solo sobrante',
        RolloverMode.full => 'Todo el saldo',
      };
}

/// Período presupuestario (zero-based). Espejo 1:1 de `public.budget_periods`.
@freezed
abstract class BudgetPeriod with _$BudgetPeriod {
  const factory BudgetPeriod({
    required String id,
    @JsonKey(name: 'household_id') required String householdId,
    @JsonKey(name: 'period_type') required PeriodType periodType,
    @JsonKey(name: 'period_start') required DateTime periodStart,
    @JsonKey(name: 'period_end') required DateTime periodEnd,
    @JsonKey(name: 'total_income')
    @DecimalConverter()
    required Decimal totalIncome,
    @JsonKey(name: 'total_allocated')
    @DecimalConverter()
    required Decimal totalAllocated,
    @JsonKey(name: 'ready_to_assign')
    @DecimalConverter()
    required Decimal readyToAssign,
    String? notes,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _BudgetPeriod;

  factory BudgetPeriod.fromJson(Map<String, dynamic> json) =>
      _$BudgetPeriodFromJson(json);
}

/// Asignación a un envelope dentro de un período. Espejo 1:1 de
/// `public.budget_allocations`.
@freezed
abstract class BudgetAllocation with _$BudgetAllocation {
  const factory BudgetAllocation({
    required String id,
    @JsonKey(name: 'period_id') required String periodId,
    required String category,
    required String subcategory,
    @DecimalConverter() required Decimal allocated,
    @JsonKey(name: 'rollover_from_prev')
    @DecimalConverter()
    required Decimal rolloverFromPrev,
    @JsonKey(name: 'rollover_mode') required RolloverMode rolloverMode,
    required String currency,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _BudgetAllocation;

  factory BudgetAllocation.fromJson(Map<String, dynamic> json) =>
      _$BudgetAllocationFromJson(json);
}

/// Cálculo ligero del saldo de un envelope en cliente. La fuente canónica
/// sigue siendo la función SQL `public.envelope_balance(...)`. No serializa
/// (no es una tabla) — espejo de `EnvelopeStatus` en iOS.
@freezed
abstract class EnvelopeStatus with _$EnvelopeStatus {
  const EnvelopeStatus._();

  const factory EnvelopeStatus({
    required String category,
    required String subcategory,
    @DecimalConverter() required Decimal allocated,
    @DecimalConverter() required Decimal spent,
  }) = _EnvelopeStatus;

  /// Saldo restante del envelope.
  Decimal get remaining => allocated - spent;

  /// Porcentaje usado, clampeado a [0, 1].
  double get percentUsed {
    if (allocated <= Decimal.zero) return 0;
    final double ratio = spent.toDouble() / allocated.toDouble();
    return ratio.clamp(0, 1).toDouble();
  }

  /// Indica sobregiro del envelope.
  bool get isOverBudget => spent > allocated;
}
